demos/device_defender_for_aws/defender_demo.c

```diff
@@ -0,0 +1,697 @@
+         * current timestamp. */
+        reportId = ( uint32_t ) xTaskGetTickCount();
+
+        LogInfo( ( "Establishing MQTT session..." ) );
+        demoStatus = EstablishMqttSession( &mqttContext,
+                                           &networkContext,
+                                           &xBuffer,
+                                           publishCallback );
+
+        if( demoStatus == pdFAIL )
+        {
+            LogError( ( "Failed to establish MQTT session." ) );
+        }
+        else
+        {
+            mqttSessionEstablished = 1;
+        }
+
+        if( demoStatus == pdPASS )
+        {
+            LogInfo( ( "Subscribing to defender topics..." ) );
+            demoStatus = subscribeToDefenderTopics( &mqttContext );
+
+            if( demoStatus == pdFAIL )
+            {
+                LogError( ( "Failed to subscribe to defender topics." ) );
+            }
+        }
+
+        if( demoStatus == pdPASS )
+        {
+            LogInfo( ( "Collecting device metrics..." ) );
+            demoStatus = collectDeviceMetrics();
+
+            if( demoStatus == pdFAIL )
+            {
+                LogError( ( "Failed to collect device metrics." ) );
+            }
+        }
+
+        if( demoStatus == pdPASS )
+        {
+            LogInfo( ( "Generating device defender report..." ) );
+            demoStatus = generateDeviceMetricsReport( &( reportLength ) );
+
+            if( demoStatus == pdFAIL )
+            {
+                LogError( ( "Failed to generate device defender report." ) );
+            }
+        }
+
+        if( demoStatus == pdPASS )
+        {
+            LogInfo( ( "Publishing device defender report..." ) );
+            demoStatus = PublishToTopic( &mqttContext,
+                                         DEFENDER_API_JSON_PUBLISH( THING_NAME ),
+                                         DEFENDER_API_LENGTH_JSON_PUBLISH( THING_NAME_LENGTH ),
+                                         &( deviceMetricsJsonReport[ 0 ] ),
+                                         reportLength );
+
+            if( demoStatus == pdFAIL )
+            {
+                LogError( ( "Failed to publish device defender report." ) );
+            }
+        }
+
+        if( demoStatus == pdPASS )
+        {
+            /* Note that PublishToTopic already called MQTT_ProcessLoop, therefore
+             * responses may have been received and the prvEventCallback may have
+             * been called. */
+            for( i = 0; i < DEFENDER_RESPONSE_WAIT_SECONDS; i++ )
+            {
+                /* reportStatus is updated in the publishCallback. */
+                if( reportStatus != ReportStatusNotReceived )
+                {
+                    break;
+                }
+
+                /* Wait for sometime between consecutive executions of ProcessLoop. */
+                vTaskDelay( pdMS_TO_TICKS( 1000 ) );
+
+                ( void ) ProcessLoop( &mqttContext );
```

Comment:
Why not add a `uint32_t` parameter to `ProcessLoop` and call it with 1000?

Reply:
Since `vTaskDelay` is called with 1000 here, and ProcessLoop() waits for `mqttexamplePROCESS_LOOP_TIMEOUT_MS` or 500ms, the total wait here is `DEFENDER_RESPONSE_WAIT_SECONDS` * 1.5, which is misleading